### PR TITLE
[Rust Frontend] Support prompt-only completions

### PR DESCRIPTION
**Pull request by reidliu41**
## Purpose

Part of https://github.com/vllm-project/vllm/issues/44280.

  The Rust OpenAI completions route rejected `max_tokens=0` unconditionally. Python vLLM supports the prompt-only completions case
  when `echo=true`, where the API returns the prompt without exposing any generated completion tokens.

  This PR adds that Rust compatibility path for `/v1/completions`:

  - Allows `echo=true` with `max_tokens=0`.
  - Rejects `max_tokens=0` when `echo=false` with a clear validation error.
  - Lowers prompt-only requests to one internal token so the engine path can run.
  - Hides the internal token from text, token IDs, and usage.
  - Preserves prompt-only behavior for non-streaming, streaming, and logprobs responses.

## Test 

```
cargo test -p vllm-server completions
test result: ok. 91 passed; 0 failed; 0 ignored; 0 measured; 98 filtered out; finished in 0.48s

cargo test -p vllm-server
  test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

 cargo clippy -p vllm-server --tests -- -D warnings
  Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.15s

 cargo test --workspace -- --test-threads=1
test result: ok. 251 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s



  cargo run --bin vllm-rs -- serve Qwen/Qwen3-0.6B \
    --port 18081 \
    --max-model-len 512 \
    --served-model-name m \
    --engine-ready-timeout-secs 240 \
    -- \
    --dtype float16 \
    --enforce-eager \
    --gpu-memory-utilization 0.3



$   curl -sS http://127.0.0.1:18081/v1/completions \
>     -H 'Content-Type: application/json' \
>     -d '{"model":"m","prompt":"Hello","echo":true,"max_tokens":0,"return_token_ids":true}'
{"id":"cmpl-be81a79c","object":"text_completion","created":1780960254,"model":"m","choices":[{"index":0,"text":"Hello","finish_reason":"length","token_ids":[],"prompt_token_ids":[9707]}],"usage":{"prompt_tokens":1,"total_tokens":1,"completion_tokens":0}}


$   curl -sS http://127.0.0.1:18081/v1/completions \
>     -H 'Content-Type: application/json' \
>     -d '{"model":"m","prompt":"Hello","echo":false,"max_tokens":0}'
{"error":{"message":"max_tokens=0 is only supported when echo=true.","type":"invalid_request_error","param":"max_tokens","code":"invalid_request_error"}}



$   curl -sS -N http://127.0.0.1:18081/v1/completions \
>     -H 'Content-Type: application/json' \
>     -d '{"model":"m","prompt":"Hello","echo":true,"max_tokens":0,"stream":true,"stream_options":
>     {"include_usage":true},"return_token_ids":true}'
data: {"id":"cmpl-37f7af70","object":"text_completion","created":1780960304,"model":"m","choices":[{"index":0,"text":"Hello","prompt_token_ids":[9707]}]}

data: {"id":"cmpl-37f7af70","object":"text_completion","created":1780960304,"model":"m","choices":[{"index":0,"text":"","finish_reason":"length"}]}

data: {"id":"cmpl-37f7af70","object":"text_completion","created":1780960304,"model":"m","choices":[],"usage":{"prompt_tokens":1,"total_tokens":1,"completion_tokens":0}}

data: [DONE]


$   curl -sS http://127.0.0.1:18081/v1/completions \
>     -H 'Content-Type: application/json' \
>     -d '{"model":"m","prompt":"Hello","echo":true,"max_tokens":0,"logprobs":1}'
{"id":"cmpl-0bf73284","object":"text_completion","created":1780960356,"model":"m","choices":[{"index":0,"text":"Hello","logprobs":{"tokens":["Hello"],"token_logprobs":[null],"top_logprobs":[null],"text_offset":[0]},"finish_reason":"length","prompt_logprobs":[null]}],"usage":{"prompt_tokens":1,"total_tokens":1,"completion_tokens":0}}

```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: rust/src/server/src/routes/openai/completions.rs [hunk 1]
```diff
@@ -188,12 +234,18 @@ async fn collect_completion(
...
+                } else {
+                    collected.token_ids.clone()
+                }
+            }),
             prompt_token_ids: return_token_ids.then(|| collected.prompt_token_ids.to_vec()),
         }],
         usage: Some(Usage::from_counts(
             collected.prompt_token_ids.len() as u32,
-            collected.token_ids.len() as u32,
+            completion_token_count,
```
Comment (tahsintunan) [inline rust/src/server/src/routes/openai/completions.rs]: One open question: this PR reports `completion_tokens=0`, Python reports `completion_tokens=1` (i.e. [counts](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/completion/serving.py#L573) the internal token).

Case for 0: it's an implementation artifact. `max_tokens=0` was requested, and `completion_tokens=1` breaks the `<= max_tokens` invariant some clients assume.

Case for 1: the decode step really ran, and billing/reconciliation pipelines _may_ currently agree with Python exactly. This would drift by one per prompt-only request.

Worth noting this rather than settling it implicitly here.

cc @BugenZhao

File: rust/src/server/src/routes/openai/completions.rs [hunk 2]
```diff
@@ -149,17 +152,34 @@ async fn collect_completion(
...
+                server_error!(
+                    "prompt-only completion requested logprobs but generation returned none"
+                )
+            })?;
+            Some(decoded_prompt_logprobs_to_openai(
+                prompt_logprobs,
+                0,
+                return_tokens_as_token_ids,
+            )?)
+        }
```
Comment (tahsintunan) [inline rust/src/server/src/routes/openai/completions.rs]: The `else` arm can only ever error. If the `if let` failed, `prompt_logprobs` is `None`, so `ok_or_else` always trips and the conversion after it is unreachable. Flattening it to a plain error arm would make that explicit.

Also, this block and the streaming `Start` one are the same logic, so a shared helper would deduplicate them. Though if the fallback turns out to be dead (see my other comment), these may shrink enough that flattening alone is cleaner.

File: rust/src/server/src/routes/openai/completions.rs [hunk 3]
```diff
@@ -336,6 +455,23 @@ fn delta_chunk(
...
+            format!("token_id:{token_id}")
+        } else {
+            prompt.to_string()
+        }],
+        token_logprobs: vec![None],
+        top_logprobs: vec![None],
+        text_offset: vec![0],
+    }
+}
+
```
Comment (tahsintunan) [inline rust/src/server/src/routes/openai/completions.rs]: Can you double-check whether this fallback is reachable?

From my understanding, a 1-token prompt with `prompt_logprobs` requested still produces an (empty) payload: the worker inserts it before the `num_logits <= 0` continue, the scheduler attaches it unconditionally, and `decode_prompt_logprobs` builds `first_token` from `prompt_token_ids`, not the payload.

If that's right, the route always sees `Some(scored_positions: [])`, which `decoded_prompt_logprobs_to_openai` already handles the same way. The unit tests only hit the branch because they construct `Start` events with `prompt_logprobs: None` by hand. If the helper is confirmed dead, it and the flag can go, and the blocks reduce to `Some -> convert, None -> server error`.

**Comment by reidliu41** [comment]
Thanks for your review. 
- Updated to match Python for now, consistent with recent PRs fixing Rust/Python frontend parity issues: prompt-only
  completions count the internal decode token, so completion_tokens is 1.
-  Done. Prompt-only logprobs handling is now shared through helpers, and the unreachable error/conversion branch was removed.
- Double checked with live HTTP testing. The single-token fallback is reachable because the engine can return no
  prompt_logprobs for that path, so removing it causes a server error. Kept the fallback and added coverage.

**Comment by BugenZhao** [comment]
@codex review

File: rust/src/server/src/routes/openai/completions.rs [hunk 4]
```diff
@@ -163,10 +168,18 @@ async fn collect_completion(
...
+            collected.prompt_logprobs.as_ref(),
+            collected.prompt_token_ids.as_ref(),
+            return_tokens_as_token_ids,
+        )?)
+    } else {
+        None
+    };
     let text = match &echo {
         None => collected.text,
+        Some(prompt) if prompt_only => prompt.clone(),
```
[the remark this answers is not in the document]
Comment (reidliu41) [reply rust/src/server/src/routes/openai/completions.rs:182]: Thanks for the review. I’m keeping this unchanged to match Python vLLM behavior for now: non-streaming prompt-only completions still count the internal decode token in usage, and return_token_ids exposes output.token_ids the same way Python does.

File: rust/src/server/src/routes/openai/completions.rs [hunk 5]
```diff
@@ -278,6 +308,45 @@ async fn completion_chunk_stream(
...
+                                prompt_tokens = finished.usage.prompt_token_count,
+                                output_tokens = finished.usage.output_token_count,
+                                finish_reason = finished.finish_reason.as_str(),
+                                "completion finished"
+                            );
+                        }
+                        continuous_usage.set_final_counts(
+                            finished.usage.prompt_token_count,
+                            finished.usage.output_token_count,
+                        );
```
[the remark this answers is not in the document]
Comment (reidliu41) [reply rust/src/server/src/routes/openai/completions.rs:329]: Thanks for the review. I’m keeping this unchanged to match Python vLLM behavior for now: streaming prompt-only completions also count the internal decode token in continuous usage and the final usage chunk, consistent with Python.

File: rust/src/server/src/routes/openai/completions.rs [hunk 6]
```diff
@@ -278,6 +308,45 @@ async fn completion_chunk_stream(
                 logprobs,
                 finished,
             }) => {
+                if prompt_only {
+                    if let Some(finished) = finished {
```
Comment (BugenZhao) [inline rust/src/server/src/routes/openai/completions.rs:315]: Would you add some comments for the branching here?

**Review by BugenZhao — APPROVED** [review]
LGTM. Thanks!

File: rust/src/server/src/routes/openai/completions.rs (hunk 6 above)
Comment (reidliu41) [reply rust/src/server/src/routes/openai/completions.rs:315]: updated
